[miele] Exception on turning the hood light on and off

I am running the snapshot build 2.5.0 1663. When turning the hood light on and off via the miele binding, I get the following exception. Turning on and off the light is still working. The exception was not introduced by the current refactorings and already occured in the past. 

Maybe the exceptions are related to the fact that my hood is a special model. The gateway is reporting the following device info:

Technical type: not defined
Electronic Software ID: 0
Fabrication number: not defined

```
2019-08-18 10:41:16.131 [ERROR] [nal.common.AbstractInvocationHandler] - An error occurred while calling method 'ThingHandler.handleCommand()' on 'org.openhab.binding.miele.internal.handler.HoodHandler@7b436a33': JsonNull
java.lang.UnsupportedOperationException: JsonNull
	at com.google.gson.JsonElement.getAsString(JsonElement.java:192) ~[21:com.google.gson:2.8.2.v20180104-1110]
	at org.openhab.binding.miele.internal.handler.HoodHandler.handleCommand(HoodHandler.java:76) ~[?:?]
	at sun.reflect.GeneratedMethodAccessor110.invoke(Unknown Source) ~[?:?]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:?]
	at java.lang.reflect.Method.invoke(Method.java:498) ~[?:?]
	at org.eclipse.smarthome.core.internal.common.AbstractInvocationHandler.invokeDirect(AbstractInvocationHandler.java:152) [133:org.openhab.core:2.5.0.201908170301]
	at org.eclipse.smarthome.core.internal.common.InvocationHandlerSync.invoke(InvocationHandlerSync.java:59) [133:org.openhab.core:2.5.0.201908170301]
	at com.sun.proxy.$Proxy4844.handleCommand(Unknown Source) [?:?]
	at org.eclipse.smarthome.core.thing.internal.profiles.ProfileCallbackImpl.handleCommand(ProfileCallbackImpl.java:75) [184:org.openhab.core.thing:2.5.0.201908170303]
	at org.eclipse.smarthome.core.thing.internal.profiles.SystemDefaultProfile.onCommandFromItem(SystemDefaultProfile.java:49) [184:org.openhab.core.thing:2.5.0.201908170303]
	at sun.reflect.GeneratedMethodAccessor109.invoke(Unknown Source) ~[?:?]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:?]
	at java.lang.reflect.Method.invoke(Method.java:498) ~[?:?]
	at org.eclipse.smarthome.core.internal.common.AbstractInvocationHandler.invokeDirect(AbstractInvocationHandler.java:152) [133:org.openhab.core:2.5.0.201908170301]
	at org.eclipse.smarthome.core.internal.common.Invocation.call(Invocation.java:53) [133:org.openhab.core:2.5.0.201908170301]
	at java.util.concurrent.FutureTask.run(FutureTask.java:266) [?:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:?]
	at java.lang.Thread.run(Thread.java:748) [?:?]
```

--- a/bundles/org.openhab.binding.miele/src/main/java/org/openhab/binding/miele/internal/MieleBindingConstants.java
+++ b/bundles/org.openhab.binding.miele/src/main/java/org/openhab/binding/miele/internal/MieleBindingConstants.java
@@ -20,6 +20,7 @@ import org.eclipse.smarthome.core.thing.ThingTypeUID;
  * used across the whole binding.
  *
  * @author Karel Goderis - Initial contribution
+ * @author Martin Lepsy - added constants for support of WiFi devices & protocol
  */
 @NonNullByDefault
 public class MieleBindingConstants {
@@ -27,6 +28,9 @@ public class MieleBindingConstants {
     public static final String BINDING_ID = "miele";
     public static final String APPLIANCE_ID = "uid";
     public static final String DEVICE_CLASS = "dc";
+    public static final String HDM_LAN = "hdm:LAN:";
+    public static final String HDM_ZIGBEE = "hdm:ZigBee:";
+    public static final String PROTOCOL_PROPERTY_NAME = "protocol";
 
     // List of all Thing Type UIDs
     public static final ThingTypeUID THING_TYPE_XGW3000 = new ThingTypeUID(BINDING_ID, "xgw3000");

--- a/bundles/org.openhab.binding.miele/src/main/java/org/openhab/binding/miele/internal/discovery/MieleApplianceDiscoveryService.java
+++ b/bundles/org.openhab.binding.miele/src/main/java/org/openhab/binding/miele/internal/discovery/MieleApplianceDiscoveryService.java
@@ -47,7 +47,7 @@ import com.google.gson.JsonElement;
 public class MieleApplianceDiscoveryService extends AbstractDiscoveryService implements ApplianceStatusListener {
 
     private final Logger logger = LoggerFactory.getLogger(MieleApplianceDiscoveryService.class);
-        
+
     private static final int SEARCH_TIME = 60;
 
     private MieleBridgeHandler mieleBridgeHandler;
@@ -93,15 +93,15 @@ public class MieleApplianceDiscoveryService extends AbstractDiscoveryService imp
         onApplianceAddedInternal(appliance);
     }
 
-    private void onApplianceAddedInternal(HomeDevice appliance) { 
+    private void onApplianceAddedInternal(HomeDevice appliance) {
         ThingUID thingUID = getThingUID(appliance);
         if (thingUID != null) {
             ThingUID bridgeUID = mieleBridgeHandler.getThing().getUID();
-            Map<String, Object> properties = new HashMap<>(2);                      
+            Map<String, Object> properties = new HashMap<>(2);
 
             properties.put(PROTOCOL_PROPERTY_NAME, appliance.getProtocol());
             properties.put(APPLIANCE_ID, appliance.getApplianceId());
-           
+
             for (JsonElement dc : appliance.DeviceClasses) {
                 if (dc.getAsString().contains("com.miele.xgw3000.gateway.hdm.deviceclasses.Miele")
                         && !dc.getAsString().equals("com.miele.xgw3000.gateway.hdm.deviceclasses.MieleAppliance")) {

--- a/bundles/org.openhab.binding.miele/src/main/java/org/openhab/binding/miele/internal/discovery/MieleApplianceDiscoveryService.java
+++ b/bundles/org.openhab.binding.miele/src/main/java/org/openhab/binding/miele/internal/discovery/MieleApplianceDiscoveryService.java
@@ -42,11 +42,12 @@ import com.google.gson.JsonElement;
  * associated with the Miele@Home gateway
  *
  * @author Karel Goderis - Initial contribution
+ * @author Martin Lepsy - Added protocol information in order so support WiFi devices
  */
 public class MieleApplianceDiscoveryService extends AbstractDiscoveryService implements ApplianceStatusListener {
 
     private final Logger logger = LoggerFactory.getLogger(MieleApplianceDiscoveryService.class);
-
+        
     private static final int SEARCH_TIME = 60;
 
     private MieleBridgeHandler mieleBridgeHandler;
@@ -92,13 +93,15 @@ public class MieleApplianceDiscoveryService extends AbstractDiscoveryService imp
         onApplianceAddedInternal(appliance);
     }
 
-    private void onApplianceAddedInternal(HomeDevice appliance) {
+    private void onApplianceAddedInternal(HomeDevice appliance) { 
         ThingUID thingUID = getThingUID(appliance);
         if (thingUID != null) {
             ThingUID bridgeUID = mieleBridgeHandler.getThing().getUID();
-            Map<String, Object> properties = new HashMap<>(2);
-            properties.put(APPLIANCE_ID,
-                    StringUtils.right(appliance.UID, appliance.UID.length() - new String("hdm:ZigBee:").length()));
+            Map<String, Object> properties = new HashMap<>(2);                      
+
+            properties.put(PROTOCOL_PROPERTY_NAME, appliance.getProtocol());
+            properties.put(APPLIANCE_ID, appliance.getApplianceId());
+           
             for (JsonElement dc : appliance.DeviceClasses) {
                 if (dc.getAsString().contains("com.miele.xgw3000.gateway.hdm.deviceclasses.Miele")
                         && !dc.getAsString().equals("com.miele.xgw3000.gateway.hdm.deviceclasses.MieleAppliance")) {
@@ -155,7 +158,7 @@ public class MieleApplianceDiscoveryService extends AbstractDiscoveryService imp
                     StringUtils.lowerCase(modelID.replaceAll("[^a-zA-Z0-9_]", "_")));
 
             if (getSupportedThingTypes().contains(thingTypeUID)) {
-                ThingUID thingUID = new ThingUID(thingTypeUID, bridgeUID, getId(appliance));
+                ThingUID thingUID = new ThingUID(thingTypeUID, bridgeUID, appliance.getId());
                 return thingUID;
             } else {
                 return null;
@@ -164,10 +167,4 @@ public class MieleApplianceDiscoveryService extends AbstractDiscoveryService imp
             return null;
         }
     }
-
-    private String getId(HomeDevice appliance) {
-        return StringUtils.right(appliance.UID, appliance.UID.length() - new String("hdm:ZigBee:").length())
-                .replaceAll("[^a-zA-Z0-9_]", "_");
-    }
-
 }

--- a/bundles/org.openhab.binding.miele/src/main/java/org/openhab/binding/miele/internal/discovery/MieleMDNSDiscoveryParticipant.java
+++ b/bundles/org.openhab.binding.miele/src/main/java/org/openhab/binding/miele/internal/discovery/MieleMDNSDiscoveryParticipant.java
@@ -61,7 +61,7 @@ public class MieleMDNSDiscoveryParticipant implements MDNSDiscoveryParticipant {
     @Override
     public DiscoveryResult createResult(ServiceInfo service) {
         if (isMieleGateway(service)) {
-            ThingUID uid = getThingUID(service); // HIER
+            ThingUID uid = getThingUID(service);
 
             if (uid != null) {
                 Map<String, Object> properties = new HashMap<>(2);
@@ -86,7 +86,7 @@ public class MieleMDNSDiscoveryParticipant implements MDNSDiscoveryParticipant {
             }
         }
         return null;
-    }    
+    }
 
     @Override
     public ThingUID getThingUID(ServiceInfo service) {
@@ -102,9 +102,12 @@ public class MieleMDNSDiscoveryParticipant implements MDNSDiscoveryParticipant {
 
     /**
      * Checks if service is a Miele XGW3000 Gateway
-     * 
+     *
      * application must be mieleathome
      * must contain path with value /rest/
+     *
+     * @param service the service to check
+     * @return true, if the discoverd service is a Miele XGW3000 Gateway
      */
     private boolean isMieleGateway(ServiceInfo service) {
         return service.getApplication().contains(SERVICE_NAME) && service.getPropertyString(PATH_PROPERTY_NAME) != null

--- a/bundles/org.openhab.binding.miele/src/main/java/org/openhab/binding/miele/internal/discovery/MieleMDNSDiscoveryParticipant.java
+++ b/bundles/org.openhab.binding.miele/src/main/java/org/openhab/binding/miele/internal/discovery/MieleMDNSDiscoveryParticipant.java
@@ -37,12 +37,16 @@ import org.slf4j.LoggerFactory;
  * {@link MDNSDiscoveryService}.
  *
  * @author Karel Goderis - Initial contribution
+ * @author Martin Lepsy - Added check for Miele gateway for cleaner discovery
  *
  */
 @Component(immediate = true)
 public class MieleMDNSDiscoveryParticipant implements MDNSDiscoveryParticipant {
 
     private final Logger logger = LoggerFactory.getLogger(MieleMDNSDiscoveryParticipant.class);
+    private static final String PATH_TO_CHECK_FOR_XGW3000 = "/rest/";
+    private static final String SERVICE_NAME = "mieleathome";
+    private static final String PATH_PROPERTY_NAME = "path";
 
     @Override
     public Set<ThingTypeUID> getSupportedThingTypeUIDs() {
@@ -56,8 +60,8 @@ public class MieleMDNSDiscoveryParticipant implements MDNSDiscoveryParticipant {
 
     @Override
     public DiscoveryResult createResult(ServiceInfo service) {
-        if (service.getApplication().contains("mieleathome")) {
-            ThingUID uid = getThingUID(service);
+        if (isMieleGateway(service)) {
+            ThingUID uid = getThingUID(service); // HIER
 
             if (uid != null) {
                 Map<String, Object> properties = new HashMap<>(2);
@@ -82,7 +86,7 @@ public class MieleMDNSDiscoveryParticipant implements MDNSDiscoveryParticipant {
             }
         }
         return null;
-    }
+    }    
 
     @Override
     public ThingUID getThingUID(ServiceInfo service) {
@@ -94,6 +98,17 @@ public class MieleMDNSDiscoveryParticipant implements MDNSDiscoveryParticipant {
         }
 
         return null;
+    }
+
+    /**
+     * Checks if service is a Miele XGW3000 Gateway
+     * 
+     * application must be mieleathome
+     * must contain path with value /rest/
+     */
+    private boolean isMieleGateway(ServiceInfo service) {
+        return service.getApplication().contains(SERVICE_NAME) && service.getPropertyString(PATH_PROPERTY_NAME) != null
+            && service.getPropertyString(PATH_PROPERTY_NAME).equalsIgnoreCase(PATH_TO_CHECK_FOR_XGW3000);
     }
 
 }

--- a/bundles/org.openhab.binding.miele/src/main/java/org/openhab/binding/miele/internal/handler/CoffeeMachineHandler.java
+++ b/bundles/org.openhab.binding.miele/src/main/java/org/openhab/binding/miele/internal/handler/CoffeeMachineHandler.java
@@ -14,8 +14,6 @@ package org.openhab.binding.miele.internal.handler;
 
 import static org.openhab.binding.miele.internal.MieleBindingConstants.APPLIANCE_ID;
 
-import com.google.gson.JsonElement;
-
 import org.eclipse.smarthome.core.library.types.OnOffType;
 import org.eclipse.smarthome.core.thing.ChannelUID;
 import org.eclipse.smarthome.core.thing.Thing;
@@ -23,6 +21,8 @@ import org.eclipse.smarthome.core.types.Command;
 import org.eclipse.smarthome.core.types.RefreshType;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
+
+import com.google.gson.JsonElement;
 
 /**
  * The {@link CoffeeMachineHandler} is responsible for handling commands,

--- a/bundles/org.openhab.binding.miele/src/main/java/org/openhab/binding/miele/internal/handler/CoffeeMachineHandler.java
+++ b/bundles/org.openhab.binding.miele/src/main/java/org/openhab/binding/miele/internal/handler/CoffeeMachineHandler.java
@@ -14,6 +14,8 @@ package org.openhab.binding.miele.internal.handler;
 
 import static org.openhab.binding.miele.internal.MieleBindingConstants.APPLIANCE_ID;
 
+import com.google.gson.JsonElement;
+
 import org.eclipse.smarthome.core.library.types.OnOffType;
 import org.eclipse.smarthome.core.thing.ChannelUID;
 import org.eclipse.smarthome.core.thing.Thing;
@@ -22,13 +24,12 @@ import org.eclipse.smarthome.core.types.RefreshType;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
-import com.google.gson.JsonElement;
-
 /**
  * The {@link CoffeeMachineHandler} is responsible for handling commands,
  * which are sent to one of the channels
  *
  * @author Stephan Esch - Initial contribution
+ * @author Martin Lepsy - fixed handling of empty JSON results
  */
 public class CoffeeMachineHandler extends MieleApplianceHandler<CoffeeMachineChannelSelector> {
 
@@ -68,7 +69,7 @@ public class CoffeeMachineHandler extends MieleApplianceHandler<CoffeeMachineCha
                 }
             }
             // process result
-            if (result != null) {
+            if (isResultProcessable(result)) {
                 logger.debug("Result of operation is {}", result.getAsString());
             }
         } catch (IllegalArgumentException e) {

--- a/bundles/org.openhab.binding.miele/src/main/java/org/openhab/binding/miele/internal/handler/DishWasherHandler.java
+++ b/bundles/org.openhab.binding.miele/src/main/java/org/openhab/binding/miele/internal/handler/DishWasherHandler.java
@@ -30,6 +30,7 @@ import com.google.gson.JsonElement;
  *
  * @author Karel Goderis - Initial contribution
  * @author Kai Kreuzer - fixed handling of REFRESH commands
+ * @author Martin Lepsy - fixed handling of empty JSON results
  */
 public class DishWasherHandler extends MieleApplianceHandler<DishwasherChannelSelector> {
 
@@ -69,7 +70,7 @@ public class DishWasherHandler extends MieleApplianceHandler<DishwasherChannelSe
                 }
             }
             // process result
-            if (result != null) {
+            if (isResultProcessable(result)) {
                 logger.debug("Result of operation is {}", result.getAsString());
             }
         } catch (IllegalArgumentException e) {

--- a/bundles/org.openhab.binding.miele/src/main/java/org/openhab/binding/miele/internal/handler/FridgeFreezerHandler.java
+++ b/bundles/org.openhab.binding.miele/src/main/java/org/openhab/binding/miele/internal/handler/FridgeFreezerHandler.java
@@ -14,6 +14,8 @@ package org.openhab.binding.miele.internal.handler;
 
 import static org.openhab.binding.miele.internal.MieleBindingConstants.APPLIANCE_ID;
 
+import com.google.gson.JsonElement;
+
 import org.eclipse.smarthome.core.library.types.OnOffType;
 import org.eclipse.smarthome.core.thing.ChannelUID;
 import org.eclipse.smarthome.core.thing.Thing;
@@ -21,14 +23,13 @@ import org.eclipse.smarthome.core.types.Command;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
-import com.google.gson.JsonElement;
-
 /**
  * The {@link FridgeFreezerHandler} is responsible for handling commands,
  * which are sent to one of the channels
  *
  * @author Karel Goderis - Initial contribution
  * @author Kai Kreuzer - fixed handling of REFRESH commands
+ * @author Martin Lepsy - fixed handling of empty JSON results
  */
 public class FridgeFreezerHandler extends MieleApplianceHandler<FridgeFreezerChannelSelector> {
 
@@ -74,7 +75,7 @@ public class FridgeFreezerHandler extends MieleApplianceHandler<FridgeFreezerCha
                 }
             }
             // process result
-            if (result != null) {
+            if (isResultProcessable(result)) {
                 logger.debug("Result of operation is {}", result.getAsString());
             }
         } catch (IllegalArgumentException e) {

--- a/bundles/org.openhab.binding.miele/src/main/java/org/openhab/binding/miele/internal/handler/FridgeFreezerHandler.java
+++ b/bundles/org.openhab.binding.miele/src/main/java/org/openhab/binding/miele/internal/handler/FridgeFreezerHandler.java
@@ -14,14 +14,14 @@ package org.openhab.binding.miele.internal.handler;
 
 import static org.openhab.binding.miele.internal.MieleBindingConstants.APPLIANCE_ID;
 
-import com.google.gson.JsonElement;
-
 import org.eclipse.smarthome.core.library.types.OnOffType;
 import org.eclipse.smarthome.core.thing.ChannelUID;
 import org.eclipse.smarthome.core.thing.Thing;
 import org.eclipse.smarthome.core.types.Command;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
+
+import com.google.gson.JsonElement;
 
 /**
  * The {@link FridgeFreezerHandler} is responsible for handling commands,

--- a/bundles/org.openhab.binding.miele/src/main/java/org/openhab/binding/miele/internal/handler/FridgeHandler.java
+++ b/bundles/org.openhab.binding.miele/src/main/java/org/openhab/binding/miele/internal/handler/FridgeHandler.java
@@ -14,6 +14,8 @@ package org.openhab.binding.miele.internal.handler;
 
 import static org.openhab.binding.miele.internal.MieleBindingConstants.APPLIANCE_ID;
 
+import com.google.gson.JsonElement;
+
 import org.eclipse.smarthome.core.library.types.OnOffType;
 import org.eclipse.smarthome.core.thing.ChannelUID;
 import org.eclipse.smarthome.core.thing.Thing;
@@ -22,13 +24,12 @@ import org.eclipse.smarthome.core.types.RefreshType;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
-import com.google.gson.JsonElement;
-
 /**
  * The {@link FridgeHandler} is responsible for handling commands,
  * which are sent to one of the channels
  *
  * @author Karel Goderis - Initial contribution
+ * @author Martin Lepsy - fixed handling of empty JSON results
  */
 public class FridgeHandler extends MieleApplianceHandler<FridgeChannelSelector> {
 
@@ -74,7 +75,7 @@ public class FridgeHandler extends MieleApplianceHandler<FridgeChannelSelector> 
                 }
             }
             // process result
-            if (result != null) {
+            if (isResultProcessable(result)) {
                 logger.debug("Result of operation is {}", result.getAsString());
             }
         } catch (IllegalArgumentException e) {

--- a/bundles/org.openhab.binding.miele/src/main/java/org/openhab/binding/miele/internal/handler/FridgeHandler.java
+++ b/bundles/org.openhab.binding.miele/src/main/java/org/openhab/binding/miele/internal/handler/FridgeHandler.java
@@ -14,8 +14,6 @@ package org.openhab.binding.miele.internal.handler;
 
 import static org.openhab.binding.miele.internal.MieleBindingConstants.APPLIANCE_ID;
 
-import com.google.gson.JsonElement;
-
 import org.eclipse.smarthome.core.library.types.OnOffType;
 import org.eclipse.smarthome.core.thing.ChannelUID;
 import org.eclipse.smarthome.core.thing.Thing;
@@ -23,6 +21,8 @@ import org.eclipse.smarthome.core.types.Command;
 import org.eclipse.smarthome.core.types.RefreshType;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
+
+import com.google.gson.JsonElement;
 
 /**
  * The {@link FridgeHandler} is responsible for handling commands,

--- a/bundles/org.openhab.binding.miele/src/main/java/org/openhab/binding/miele/internal/handler/HoodHandler.java
+++ b/bundles/org.openhab.binding.miele/src/main/java/org/openhab/binding/miele/internal/handler/HoodHandler.java
@@ -14,14 +14,14 @@ package org.openhab.binding.miele.internal.handler;
 
 import static org.openhab.binding.miele.internal.MieleBindingConstants.APPLIANCE_ID;
 
-import com.google.gson.JsonElement;
-
 import org.eclipse.smarthome.core.library.types.OnOffType;
 import org.eclipse.smarthome.core.thing.ChannelUID;
 import org.eclipse.smarthome.core.thing.Thing;
 import org.eclipse.smarthome.core.types.Command;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
+
+import com.google.gson.JsonElement;
 
 /**
  * The {@link HoodHandler} is responsible for handling commands,

--- a/bundles/org.openhab.binding.miele/src/main/java/org/openhab/binding/miele/internal/handler/HoodHandler.java
+++ b/bundles/org.openhab.binding.miele/src/main/java/org/openhab/binding/miele/internal/handler/HoodHandler.java
@@ -14,6 +14,8 @@ package org.openhab.binding.miele.internal.handler;
 
 import static org.openhab.binding.miele.internal.MieleBindingConstants.APPLIANCE_ID;
 
+import com.google.gson.JsonElement;
+
 import org.eclipse.smarthome.core.library.types.OnOffType;
 import org.eclipse.smarthome.core.thing.ChannelUID;
 import org.eclipse.smarthome.core.thing.Thing;
@@ -21,14 +23,13 @@ import org.eclipse.smarthome.core.types.Command;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
-import com.google.gson.JsonElement;
-
 /**
  * The {@link HoodHandler} is responsible for handling commands,
  * which are sent to one of the channels
  *
  * @author Karel Goderis - Initial contribution
  * @author Kai Kreuzer - fixed handling of REFRESH commands
+ * @author Martin Lepsy - fixed handling of empty JSON results
  */
 public class HoodHandler extends MieleApplianceHandler<HoodChannelSelector> {
 
@@ -72,7 +73,7 @@ public class HoodHandler extends MieleApplianceHandler<HoodChannelSelector> {
                 }
             }
             // process result
-            if (result != null) {
+            if (isResultProcessable(result)) {
                 logger.debug("Result of operation is {}", result.getAsString());
             }
         } catch (IllegalArgumentException e) {

--- a/bundles/org.openhab.binding.miele/src/main/java/org/openhab/binding/miele/internal/handler/MieleApplianceHandler.java
+++ b/bundles/org.openhab.binding.miele/src/main/java/org/openhab/binding/miele/internal/handler/MieleApplianceHandler.java
@@ -52,6 +52,7 @@ import com.google.gson.JsonParser;
  * the {@link ApplianceChannelSelector} datapoints
  *
  * @author Karel Goderis - Initial contribution
+ * @author Martin Lepsy - Added check for JsonNull result
  */
 public abstract class MieleApplianceHandler<E extends Enum<E> & ApplianceChannelSelector> extends BaseThingHandler
         implements ApplianceStatusListener {
@@ -142,8 +143,8 @@ public abstract class MieleApplianceHandler<E extends Enum<E> & ApplianceChannel
     }
 
     @Override
-    public void onApplianceStateChanged(String UID, DeviceClassObject dco) {
-        String myUID = "hdm:ZigBee:" + (String) getThing().getConfiguration().getProperties().get(APPLIANCE_ID);
+    public void onApplianceStateChanged(String UID, DeviceClassObject dco) { 
+        String myUID = ((String) getThing().getProperties().get(PROTOCOL_PROPERTY_NAME))  + (String) getThing().getConfiguration().getProperties().get(APPLIANCE_ID);
         String modelID = StringUtils.right(dco.DeviceClass,
                 dco.DeviceClass.length() - new String("com.miele.xgw3000.gateway.hdm.deviceclasses.Miele").length());
 
@@ -175,7 +176,7 @@ public abstract class MieleApplianceHandler<E extends Enum<E> & ApplianceChannel
 
     @Override
     public void onAppliancePropertyChanged(String UID, DeviceProperty dp) {
-        String myUID = "hdm:ZigBee:" + (String) getThing().getConfiguration().getProperties().get(APPLIANCE_ID);
+        String myUID = ((String) getThing().getProperties().get(PROTOCOL_PROPERTY_NAME)) + (String) getThing().getConfiguration().getProperties().get(APPLIANCE_ID);
 
         if (myUID.equals(UID)) {
             try {
@@ -269,5 +270,9 @@ public abstract class MieleApplianceHandler<E extends Enum<E> & ApplianceChannel
             }
         }
         return this.bridgeHandler;
+    }
+
+    protected boolean isResultProcessable(JsonElement result) {
+        return result != null && !result.isJsonNull();
     }
 }

--- a/bundles/org.openhab.binding.miele/src/main/java/org/openhab/binding/miele/internal/handler/MieleBridgeHandler.java
+++ b/bundles/org.openhab.binding.miele/src/main/java/org/openhab/binding/miele/internal/handler/MieleBridgeHandler.java
@@ -97,7 +97,7 @@ public class MieleBridgeHandler extends BaseBridgeHandler {
 
     // Data structures to de-JSONify whatever Miele appliances are sending us
     public class HomeDevice {
-     
+
         private static final String PROTOCOL_LAN = "LAN";
 
         public String Name;
@@ -117,7 +117,7 @@ public class MieleBridgeHandler extends BaseBridgeHandler {
         }
 
         public String getId() {
-            return getApplianceId().replaceAll("[^a-zA-Z0-9_]", "_");           
+            return getApplianceId().replaceAll("[^a-zA-Z0-9_]", "_");
         }
 
         public String getProtocol() {
@@ -263,8 +263,8 @@ public class MieleBridgeHandler extends BaseBridgeHandler {
                             previousHomeDevices = currentHomeDevices;
 
                             for (Thing appliance : getThing().getThings()) {
-                                if (appliance.getStatus() == ThingStatus.ONLINE) {                                                                        
-                                    String UID = appliance.getProperties().get(PROTOCOL_PROPERTY_NAME) 
+                                if (appliance.getStatus() == ThingStatus.ONLINE) {
+                                    String UID = appliance.getProperties().get(PROTOCOL_PROPERTY_NAME)
                                             + (String) appliance.getConfiguration().getProperties().get(APPLIANCE_ID);
 
                                     Object[] args = new Object[2];
@@ -428,11 +428,11 @@ public class MieleBridgeHandler extends BaseBridgeHandler {
     };
 
     public JsonElement invokeOperation(String UID, String modelID, String methodName) {
-        return invokeOperation(UID, modelID, methodName, "hdm:ZigBee:");      
+        return invokeOperation(UID, modelID, methodName, HDM_ZIGBEE);
     }
 
     public JsonElement invokeOperation(String UID, String modelID, String methodName, String protocol) {
-        
+
         if (getThing().getStatus() == ThingStatus.ONLINE) {
             Object[] args = new Object[4];
             args[0] = protocol + UID;

--- a/bundles/org.openhab.binding.miele/src/main/java/org/openhab/binding/miele/internal/handler/MieleBridgeHandler.java
+++ b/bundles/org.openhab.binding.miele/src/main/java/org/openhab/binding/miele/internal/handler/MieleBridgeHandler.java
@@ -66,6 +66,7 @@ import com.google.gson.JsonParser;
  *
  * @author Karel Goderis - Initial contribution
  * @author Kai Kreuzer - Fixed lifecycle issues
+ * @author Martin Lepsy - Added protocol information to support WiFi devices & some refactoring for HomeDevice
  */
 public class MieleBridgeHandler extends BaseBridgeHandler {
 
@@ -96,6 +97,9 @@ public class MieleBridgeHandler extends BaseBridgeHandler {
 
     // Data structures to de-JSONify whatever Miele appliances are sending us
     public class HomeDevice {
+     
+        private static final String PROTOCOL_LAN = "LAN";
+
         public String Name;
         public String Status;
         public String ParentUID;
@@ -110,6 +114,19 @@ public class MieleBridgeHandler extends BaseBridgeHandler {
         public JsonObject Properties;
 
         HomeDevice() {
+        }
+
+        public String getId() {
+            return getApplianceId().replaceAll("[^a-zA-Z0-9_]", "_");           
+        }
+
+        public String getProtocol() {
+            return ProtocolAdapterName.equals(PROTOCOL_LAN) ? HDM_LAN : HDM_ZIGBEE;
+        }
+
+        public String getApplianceId() {
+            return ProtocolAdapterName.equals(PROTOCOL_LAN) ? StringUtils.right(UID, UID.length() - HDM_LAN.length())
+                : StringUtils.right(UID, UID.length() - HDM_ZIGBEE.length());
         }
     }
 
@@ -246,8 +263,8 @@ public class MieleBridgeHandler extends BaseBridgeHandler {
                             previousHomeDevices = currentHomeDevices;
 
                             for (Thing appliance : getThing().getThings()) {
-                                if (appliance.getStatus() == ThingStatus.ONLINE) {
-                                    String UID = "hdm:ZigBee:"
+                                if (appliance.getStatus() == ThingStatus.ONLINE) {                                                                        
+                                    String UID = appliance.getProperties().get(PROTOCOL_PROPERTY_NAME) 
                                             + (String) appliance.getConfiguration().getProperties().get(APPLIANCE_ID);
 
                                     Object[] args = new Object[2];
@@ -411,9 +428,14 @@ public class MieleBridgeHandler extends BaseBridgeHandler {
     };
 
     public JsonElement invokeOperation(String UID, String modelID, String methodName) {
+        return invokeOperation(UID, modelID, methodName, "hdm:ZigBee:");      
+    }
+
+    public JsonElement invokeOperation(String UID, String modelID, String methodName, String protocol) {
+        
         if (getThing().getStatus() == ThingStatus.ONLINE) {
             Object[] args = new Object[4];
-            args[0] = "hdm:ZigBee:" + UID;
+            args[0] = protocol + UID;
             args[1] = "com.miele.xgw3000.gateway.hdm.deviceclasses.Miele" + modelID;
             args[2] = methodName;
             args[3] = null;
@@ -423,6 +445,7 @@ public class MieleBridgeHandler extends BaseBridgeHandler {
             return null;
         }
     }
+
 
     protected JsonElement invokeRPC(String methodName, Object[] args) {
         int id = rand.nextInt(Integer.MAX_VALUE);

--- a/bundles/org.openhab.binding.miele/src/main/java/org/openhab/binding/miele/internal/handler/OvenHandler.java
+++ b/bundles/org.openhab.binding.miele/src/main/java/org/openhab/binding/miele/internal/handler/OvenHandler.java
@@ -13,6 +13,9 @@
 package org.openhab.binding.miele.internal.handler;
 
 import static org.openhab.binding.miele.internal.MieleBindingConstants.APPLIANCE_ID;
+import static org.openhab.binding.miele.internal.MieleBindingConstants.PROTOCOL_PROPERTY_NAME;
+
+import com.google.gson.JsonElement;
 
 import org.eclipse.smarthome.core.library.types.OnOffType;
 import org.eclipse.smarthome.core.thing.ChannelUID;
@@ -22,14 +25,13 @@ import org.eclipse.smarthome.core.types.RefreshType;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
-import com.google.gson.JsonElement;
-
 /**
  * The {@link OvenHandler} is responsible for handling commands,
  * which are sent to one of the channels
  *
  * @author Karel Goderis - Initial contribution
  * @author Kai Kreuzer - fixed handling of REFRESH commands
+ * @author Martin Lepsy - fixed handling of empty JSON results
  */
 public class OvenHandler extends MieleApplianceHandler<OvenChannelSelector> {
 
@@ -45,6 +47,7 @@ public class OvenHandler extends MieleApplianceHandler<OvenChannelSelector> {
 
         String channelID = channelUID.getId();
         String uid = (String) getThing().getConfiguration().getProperties().get(APPLIANCE_ID);
+        String protocol = (String) getThing().getProperties().get(PROTOCOL_PROPERTY_NAME);
 
         OvenChannelSelector selector = (OvenChannelSelector) getValueSelectorFromChannelID(channelID);
         JsonElement result = null;
@@ -54,15 +57,15 @@ public class OvenHandler extends MieleApplianceHandler<OvenChannelSelector> {
                 switch (selector) {
                     case SWITCH: {
                         if (command.equals(OnOffType.ON)) {
-                            result = bridgeHandler.invokeOperation(uid, modelID, "switchOn");
+                            result = bridgeHandler.invokeOperation(uid, modelID, "switchOn", protocol);
                         } else if (command.equals(OnOffType.OFF)) {
-                            result = bridgeHandler.invokeOperation(uid, modelID, "switchOff");
+                            result = bridgeHandler.invokeOperation(uid, modelID, "switchOff", protocol);
                         }
                         break;
                     }
                     case STOP: {
                         if (command.equals(OnOffType.ON)) {
-                            result = bridgeHandler.invokeOperation(uid, modelID, "stop");
+                            result = bridgeHandler.invokeOperation(uid, modelID, "stop", protocol);
                         }
                         break;
                     }
@@ -75,7 +78,7 @@ public class OvenHandler extends MieleApplianceHandler<OvenChannelSelector> {
                 }
             }
             // process result
-            if (result != null) {
+            if (isResultProcessable(result)) {
                 logger.debug("Result of operation is {}", result.getAsString());
             }
         } catch (IllegalArgumentException e) {

--- a/bundles/org.openhab.binding.miele/src/main/java/org/openhab/binding/miele/internal/handler/OvenHandler.java
+++ b/bundles/org.openhab.binding.miele/src/main/java/org/openhab/binding/miele/internal/handler/OvenHandler.java
@@ -15,8 +15,6 @@ package org.openhab.binding.miele.internal.handler;
 import static org.openhab.binding.miele.internal.MieleBindingConstants.APPLIANCE_ID;
 import static org.openhab.binding.miele.internal.MieleBindingConstants.PROTOCOL_PROPERTY_NAME;
 
-import com.google.gson.JsonElement;
-
 import org.eclipse.smarthome.core.library.types.OnOffType;
 import org.eclipse.smarthome.core.thing.ChannelUID;
 import org.eclipse.smarthome.core.thing.Thing;
@@ -24,6 +22,8 @@ import org.eclipse.smarthome.core.types.Command;
 import org.eclipse.smarthome.core.types.RefreshType;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
+
+import com.google.gson.JsonElement;
 
 /**
  * The {@link OvenHandler} is responsible for handling commands,

--- a/bundles/org.openhab.binding.miele/src/main/java/org/openhab/binding/miele/internal/handler/TumbleDryerHandler.java
+++ b/bundles/org.openhab.binding.miele/src/main/java/org/openhab/binding/miele/internal/handler/TumbleDryerHandler.java
@@ -14,6 +14,8 @@ package org.openhab.binding.miele.internal.handler;
 
 import static org.openhab.binding.miele.internal.MieleBindingConstants.APPLIANCE_ID;
 
+import com.google.gson.JsonElement;
+
 import org.eclipse.smarthome.core.library.types.OnOffType;
 import org.eclipse.smarthome.core.thing.ChannelUID;
 import org.eclipse.smarthome.core.thing.Thing;
@@ -22,14 +24,13 @@ import org.eclipse.smarthome.core.types.RefreshType;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
-import com.google.gson.JsonElement;
-
 /**
  * The {@link TumbleDryerHandler} is responsible for handling commands,
  * which are sent to one of the channels
  *
  * @author Karel Goderis - Initial contribution
  * @author Kai Kreuzer - fixed handling of REFRESH commands
+ * @author Martin Lepsy - fixed handling of empty JSON results
  */
 public class TumbleDryerHandler extends MieleApplianceHandler<TumbleDryerChannelSelector> {
 
@@ -69,7 +70,7 @@ public class TumbleDryerHandler extends MieleApplianceHandler<TumbleDryerChannel
                 }
             }
             // process result
-            if (result != null) {
+            if (isResultProcessable(result)) {
                 logger.debug("Result of operation is {}", result.getAsString());
             }
         } catch (IllegalArgumentException e) {

--- a/bundles/org.openhab.binding.miele/src/main/java/org/openhab/binding/miele/internal/handler/TumbleDryerHandler.java
+++ b/bundles/org.openhab.binding.miele/src/main/java/org/openhab/binding/miele/internal/handler/TumbleDryerHandler.java
@@ -14,8 +14,6 @@ package org.openhab.binding.miele.internal.handler;
 
 import static org.openhab.binding.miele.internal.MieleBindingConstants.APPLIANCE_ID;
 
-import com.google.gson.JsonElement;
-
 import org.eclipse.smarthome.core.library.types.OnOffType;
 import org.eclipse.smarthome.core.thing.ChannelUID;
 import org.eclipse.smarthome.core.thing.Thing;
@@ -23,6 +21,8 @@ import org.eclipse.smarthome.core.types.Command;
 import org.eclipse.smarthome.core.types.RefreshType;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
+
+import com.google.gson.JsonElement;
 
 /**
  * The {@link TumbleDryerHandler} is responsible for handling commands,

--- a/bundles/org.openhab.binding.miele/src/main/java/org/openhab/binding/miele/internal/handler/WashingMachineHandler.java
+++ b/bundles/org.openhab.binding.miele/src/main/java/org/openhab/binding/miele/internal/handler/WashingMachineHandler.java
@@ -14,6 +14,8 @@ package org.openhab.binding.miele.internal.handler;
 
 import static org.openhab.binding.miele.internal.MieleBindingConstants.APPLIANCE_ID;
 
+import com.google.gson.JsonElement;
+
 import org.eclipse.smarthome.core.library.types.OnOffType;
 import org.eclipse.smarthome.core.thing.ChannelUID;
 import org.eclipse.smarthome.core.thing.Thing;
@@ -22,14 +24,13 @@ import org.eclipse.smarthome.core.types.RefreshType;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
-import com.google.gson.JsonElement;
-
 /**
  * The {@link WashingMachineHandler} is responsible for handling commands,
  * which are sent to one of the channels
  *
  * @author Karel Goderis - Initial contribution
  * @author Kai Kreuzer - fixed handling of REFRESH commands
+ * @author Martin Lepsy - fixed handling of empty JSON results
  */
 public class WashingMachineHandler extends MieleApplianceHandler<WashingMachineChannelSelector> {
 
@@ -70,7 +71,7 @@ public class WashingMachineHandler extends MieleApplianceHandler<WashingMachineC
                 }
             }
             // process result
-            if (result != null) {
+            if (isResultProcessable(result)) {
                 logger.debug("Result of operation is {}", result.getAsString());
             }
         } catch (IllegalArgumentException e) {

--- a/bundles/org.openhab.binding.miele/src/main/java/org/openhab/binding/miele/internal/handler/WashingMachineHandler.java
+++ b/bundles/org.openhab.binding.miele/src/main/java/org/openhab/binding/miele/internal/handler/WashingMachineHandler.java
@@ -14,8 +14,6 @@ package org.openhab.binding.miele.internal.handler;
 
 import static org.openhab.binding.miele.internal.MieleBindingConstants.APPLIANCE_ID;
 
-import com.google.gson.JsonElement;
-
 import org.eclipse.smarthome.core.library.types.OnOffType;
 import org.eclipse.smarthome.core.thing.ChannelUID;
 import org.eclipse.smarthome.core.thing.Thing;
@@ -23,6 +21,8 @@ import org.eclipse.smarthome.core.types.Command;
 import org.eclipse.smarthome.core.types.RefreshType;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
+
+import com.google.gson.JsonElement;
 
 /**
  * The {@link WashingMachineHandler} is responsible for handling commands,

